### ISSUE TITLE
Animation plugin

### ISSUE THREAD

**Issue by pardom** [koekeishiya/chunkwm#144]
I'd like to try my hand at an animation plugin. Before I go off and try implementing this (I also don't know c++), I wanted to know if the API even supports such a feature.

Here's an example from Gnome/i3

![ashamedagedbantamrooster](https://user-images.githubusercontent.com/189503/28225344-5db44b06-68a0-11e7-93ac-7da7919a6773.gif)

**Comment by pardom** [comment]
As a follow up to my comment about not knowing C++, Swift ought to be an option right? My primary language is Kotlin, which is quite similar to Swift.

**Comment by chrishoage** [comment]
The accessibility API does not natively support animating window movement. 

Any animation would have to be done by manually adjusting the window position. I imagine this would be highly inefficient and look janky

**Comment by pardom** [comment]
Hmm, again I'm not sure about the capabilities of a plugin, but what if the animation could be faked by capturing an image of the window before and after movement and animating the image?

**Comment by pardom** [comment]
So for example:

(pseudocode)
```
subscribe(notif) {
    if (notif == window_focused) {
        saveWindowImage()
    }
    else if (notif == window_moved || notif == window_resized) {
        animateWindowImage()
    }
}
```

**Comment by pardom** [comment]
Actually, probably a bit more complicated to this because multiple windows can be moved or resized with one command.

**Comment by koekeishiya** [comment]
A problem with the above approach is that windows that are not in focus can also be moved or resized.

Your best bet would be to go via the scripting addition and use private  coregraphics functions to invoke custom transforms. Prety sure I saw an example of custom animations using this in the past.

**Comment by pardom** [comment]
So could something this work?

(pseudocode)
```
subscribe(notif) {
    if (notif == window_moved || notif == window_resized) {
        newImages = createNewImages()
        animateWindows(oldImages, newImages)
        oldImages = newImages
    }
}
```

I also don't know what is meant by "the scripting addition". Excuse my ignorance.

**Comment by koekeishiya** [comment]
Sorry for the vague answer earlier. I was on my phone so didn't feel like writing much.

I'll try to keep this short.

In macOS, the only official ways to interact with windows is through the *accessibility API*. This allows us to get a window handle that allows us to move the window, set its size etc. Now this is a fairly slow API and has a fair share of problems. This is the only public API supported for interacting with windows.

Now to dig a little deeper regarding how the internal system works. By default, only the application who owns the window is able to set and manipulate its properties, however, for the Dock to be able to transform and otherwise interact with windows of OTHER applications, it needs permissions. Thus, a window require the concept of a *universal owner*  (the Dock). 

These *owners* can manipulate windows using the private low-level CoreGraphics API.
These are much more responsive than the aforementioned *acceessibility API*, and are the ones you probably want to look into to achieve smooth animations. If I'm not completely mistaken, this is how macOS does its native *minimize to Dock* animation.

To get permission to call these functions on any arbitrary window, we basically need to execute code from inside the macOS Dock application, and to do so we inject our code into the Dock process, using a [*scripting bridge*](https://developer.apple.com/library/content/documentation/Cocoa/Conceptual/ScriptingBridgeConcepts/AboutScriptingBridge/AboutScriptingBridge.html#//apple_ref/doc/uid/TP40006104-CH3-SW9).

The scripting addition I referred to above is [this one](https://github.com/koekeishiya/chwm-sa)
In short, doing these kinds of thing on macOS is not a trivial task.

**Comment by pardom** [comment]
Hah, okay. Well maybe it's a bit more than I'm willing to bite off at this point, but it would still be a great UX feature if it were to be done. Thanks for the info.

**Comment by koekeishiya** [comment]
**For future reference**

The following should be all that is necessary to create animations
```
struct MeshPoint 
{
    float x, y;
};
struct CGPointWarp
{
    MeshPoint local, global;
};
typedef uint32_t CGWindowID;
typedef int CGSConnectionID;
extern CGSConnectionID CGSMainConnectionID(void);
extern CGError CGSSetWindowTransform(CGSConnectionID cid, CGWindowID wid, CGAffineTransform transform);
extern CGError CGSSetWindowWarp(CGSConnectionID cid, CGWindowID wid, int w, int h, CGPointWarp *mesh);
```
These can be applied to windows of type `NSWindow *window = [[NSWindow alloc] ..]` for test purposes and should in theory be callable on any arbitrary window if we call through the *Dock*'s *CGSConnectionID*.

**Comment by pardom** [comment]
![giphy](https://user-images.githubusercontent.com/189503/28232125-14859f58-68bd-11e7-8d7f-ff3ec491cde7.gif)

**Comment by pardom** [comment]
Since the transparency plugin takes advantage of scripting additions, could I use that as a starting point?

**Comment by koekeishiya** [comment]
@pardom You could yes, but I would recommend just starting with getting the actual window animations up and running. You do not need special permissions for this, simply create a Cocoa application that creates one (or more ?) *NSWindow's* and start experimenting on them.

**Comment by pardom** [comment]
Ah cool. Thanks.
